File: src/java/backtracking/B_9663.java
package java.backtracking;

import java.io.*;

public class B_9663 {
    static int N;
    static int[] arr;
    static int answer = 0;
    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        N = Integer.parseInt(br.readLine());
        arr = new int[N];
        
        DFS(0);
        System.out.println(answer);
    }
    
    static void DFS(int depth) {
        if(depth == N) {
            answer++;
            return;
        }
        
        for(int i=0;i<N;i++) {
            arr[depth] = i;
            if(check(depth)) 
                DFS(depth + 1);
        }
    }
    
    static boolean check(int idx) {
        for(int i=0;i<idx;i++) 
            if(arr[i] == arr[idx] || Math.abs(i - idx) == Math.abs(arr[i] - arr[idx]))
                return false;
        return true;
    }
}

File: src/kotlin/backtracking/B_9663.kt
var n = 0
lateinit var arr: IntArray
lateinit var visit: BooleanArray
var answer = 0

fun main() = with(System.`in`.bufferedReader()) {
    n = readLine().toInt()
    arr = IntArray(n)
    visit = BooleanArray(n)
    dfs(0)
    print(answer)
}

fun dfs(depth: Int) {
    if (depth == n) {
        answer++
        return
    }

    for (i in 0..<n) {
        if (visit[i] || !check(depth, i)) {
            continue
        }
        arr[depth] = i
        visit[i] = true
        dfs(depth + 1)
        visit[i] = false
    }
}

fun check(x: Int, y: Int): Boolean {
    for (i in 0..<x) {
        if (arr[i] == y) return false
        val diff = x - i
        if (y - diff in 0..<n && arr[i] == y - diff) return false
        if (y + diff in 0..<n && arr[i] == y + diff) return false
    }
    return true
}
